aten/src/ATen/native/sparse/SparseTensorMath.cpp

```diff
@@ -261,10 +267,22 @@ SparseTensor& add_out_sparse_cpu(SparseTensor& r, const SparseTensor& t, const S
   Tensor s_values = src._values();
   r.resize_as_(src);
 
+
+  if (t_values.scalar_type() != commonDtype) {
+    t_values = t_values.to(commonDtype);
+  }
+
+  if (s_values.scalar_type() != commonDtype) {
+    s_values = s_values.to(commonDtype);
+  }
+
   if (s_values.is_contiguous() && t_values.is_contiguous()) {
     LongTensor r_indices = at::empty({sparse_dim, max_nnz}, t_indices.options());
     Tensor r_values = new_values_with_size_of(s_values, max_nnz).zero_();
-    get_sparse_impl(r)->set_indices_and_values_unsafe(r_indices, r_values);
+
+    if (commonDtype != r_values.scalar_type()) {
+      r_values = r_values.to(commonDtype);
```

Comment:
We can speed this up by allocating `r_values = torch.empty_like(r_values)`. We don't actually care about the data that is initially in the result tensor.